src/hazelweb/gui/UHCode.re

```diff
@@ -111,51 +111,27 @@ let key_handlers =
   [
     Attr.on_keypress(_ => Event.Prevent_default),
     Attr.on_keydown(evt => {
-      switch (MoveKey.of_key(Key.get_key(evt))) {
-      | Some(move_key) =>
-        prevent_stop_inject(ModelAction.MoveAction(Key(move_key)))
-      | None =>
-        switch (HazelKeyCombos.of_evt(evt)) {
-        | Some(Ctrl_Z) =>
-          if (is_mac) {
-            Event.Ignore;
-          } else {
-            prevent_stop_inject(ModelAction.Undo);
-          }
-        | Some(Meta_Z) =>
-          if (is_mac) {
-            prevent_stop_inject(ModelAction.Undo);
-          } else {
-            Event.Ignore;
-          }
-        | Some(Ctrl_Shift_Z) =>
-          if (is_mac) {
-            Event.Ignore;
-          } else {
-            prevent_stop_inject(ModelAction.Redo);
-          }
-        | Some(Meta_Shift_Z) =>
-          if (is_mac) {
-            prevent_stop_inject(ModelAction.Redo);
-          } else {
-            Event.Ignore;
-          }
-        | Some(kc) =>
-          prevent_stop_inject(
-            ModelAction.EditAction(KeyComboAction.get(cursor_info, kc)),
+      let model_action: option(ModelAction.t) = {
+        let move_key = MoveKey.of_key(Key.get_key(evt));
+        let key_combo = HazelKeyCombos.of_evt(evt);
+        let single_key = JSUtil.is_single_key(evt);
+        switch (move_key, key_combo, single_key) {
+        | (Some(move_key), _, _) => Some(MoveAction(Key(move_key)))
+        | (_, Some(key_combo), _) =>
+          KeyComboAction.get_model_action(cursor_info, key_combo, is_mac)
+        | (_, _, Some(single_key)) =>
```

Comment:
this too -- might as well have all the key handling code in one place

Reply:
I feel like there are a bunch of modules which do key handling which could be folded together... The changes I'm making in uhcode add a bunch of cases, so for simplicity I made an ad-hoc wrapper around these various modules which presents a unified key data structure. I feel most of /keyboard could be merged into a unified structure, presenting a single outward interface which maps an evt to either a unified key data structure, or even maybe directly to a modelAction.